### PR TITLE
add autodocs

### LINKED ISSUE #6648: Autogenerate docs for custom components
It would be nice to document a few things for custom components:

* How to import them and use them in a Gradio app (e.g. `from gradio_richtextbox import RichTextbox`)
* What parameters the custom component supports
* What events the custom component supports

This could be included in the Custom Components gallery that @freddyaboulton is building, as well as potentially on the pypi package itself.

### LINKED ISSUE #6684: add command to autodocument custom components
CLI command to autodocument custom components.

- generate nice docs space 
- generate readme (for pypi and github)

### PR DESCRIPTION
**Pull request by pngwn**
## Description

This is the bulk of the work for autodocumenting custom components. To test this you will need to install `gradio_paramviewer` from pip. 

I will be moving this component into gradio core, so we don't need an additional dependency but I haven't quite got round to that.

I also had to remove the readme generation out while i fixed some issues but I'll add that back very shortly.

Closes #6684. Closes #6648.

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by pngwn** [comment]
I might need help with some of those type errors, I'll swap `toml` for `tomlkit` since that is already being used.

**Comment by pngwn** [comment]
One or two tweaks to come but now reviewable.

**Comment by pngwn** [comment]
Sample markdown readme: https://gist.github.com/pngwn/d53d345e3887de357b5420581a534952

File: gradio/cli/commands/components/docs.py
Comment (freddyaboulton) [inline gradio/cli/commands/components/docs.py:1]: We need to add `from __future__ import annotations` here to support 3.8/3.9.

File: gradio/cli/commands/components/docs.py
Comment (freddyaboulton) [inline gradio/cli/commands/components/docs.py:6]: toml is only included in 3.10. We need to use `tomlkit` here for older versions. It's already installed and AKAIK `toml` is identical to `tomlkit`.

File: gradio/cli/commands/components/_docs_utils.py
Comment (freddyaboulton) [inline gradio/cli/commands/components/_docs_utils.py:306]: I think we need to do `member.__annotations__` here. 

`get_type_hints` is failing for me on python 3.9 because of the use of `|` for unions.

image

File: gradio/cli/commands/components/_docs_utils.py
Comment (freddyaboulton) [inline gradio/cli/commands/components/_docs_utils.py:22]: Ruff may not be installed in a user's system. I think we should make this optional?

Comment (pngwn) [reply gradio/cli/commands/components/_docs_utils.py:22]: good point. Or we could just add it as a dep? It really does make everything more readable.

File: gradio/cli/commands/components/docs.py
Comment (freddyaboulton) [inline gradio/cli/commands/components/docs.py:10]: I ran the docs command for all of the components - I only got two failures

- `Fallback` component (e.g. no template) - fails with `KeyError: 'return'`
- `Plot` component - fails with `KeyError: 'value'`

Haven't checked the output yet. That's next.

File: gradio/cli/commands/components/_docs_utils.py
Comment (pngwn) [reply gradio/cli/commands/components/_docs_utils.py:306]: Are these two separate issues? I'm not sure i'm understanding correctly.

If i change `hints = typing.get_type_hints(member)` to `hints = typing.get_type_hints(member.__annotations)` then it fails. Is there some specific issue with an earlier python version?

> get_type_hints is failing for me on python 3.9 because of the use of | for unions.

Is this unions in the docs code, in the component code or somewhere else?

I'll test on 3.9.

File: gradio/cli/commands/components/docs.py
Comment (pngwn) [reply gradio/cli/commands/components/docs.py:10]: Ah `Fallback` has no hints or docstrings. I'll see if we can fail more gracefully or throw an error if the output isn't helpful.

Strange behaviour for `Plot`, i think it has everything is should. Will check that.

File: gradio/cli/commands/components/_docs_utils.py
Comment (freddyaboulton) [inline gradio/cli/commands/components/_docs_utils.py:657]: Are these imported classes actually used? If not we can remove

**Review by freddyaboulton — COMMENTED** [review]
Awesome PR @pngwn ! I'm still going through this but wanted to comment on some things I've noticed - 

1. In the `space.py` the `from app import demo as app` line will hang forever unless the app file launches the demo in a `if __name__ == '__main__'` clause
2. In the README, `None` shows up for the badges if urls are not specified in the pyproject toml
image

3. I can't launch the space.py for Audio docs because there's a syntax error in the file

image

4. In the README, the types in the param table always show up as empty code blocks?

image

File: gradio/cli/commands/components/_docs_utils.py
Comment (pngwn) [reply gradio/cli/commands/components/_docs_utils.py:657]: Ah no, they won't be used.

**Comment by pngwn** [comment]
@freddyaboulton Can you should be the generated README.md, for the empty code block case?

**Comment by pngwn** [comment]
I have pushed up some fixes.

- I've change the demo generated demo file to contain the `if __name__ == '__main__'` clause, I'm not sure there is anything else we can do.
- I have done some additional cleanup on the Additional classes that we show (this was the issue with the `Audio` component not working)
- Opted for simpler type introspection in version prior to 3.10 which do not have great type introspection capabilities.
- More gracefully handled missing docstring or types (try the `Fallback` component now and it should be much nicer). Likewise plot has a missing type annotation, this should work nicely now.
- Fixed the badges
- Hide parts of the docs if there is no info to show.

File: gradio/cli/commands/components/_docs_utils.py
Comment (freddyaboulton) [inline gradio/cli/commands/components/_docs_utils.py:34]: I'm getting that `--line-length` is not a valid ruff parameter name

error: unexpected argument '--line-length' found

image

Just FYI

**Comment by freddyaboulton** [comment]
This is great @pngwn ! Thanks for making the fixes. I can confirm they're working. I noticed the following things:

1. In docs.py I think we need to import tomlkit instead of toml

2. The parameter table no longer displays default values when I use 3.10

image

If I use 3.9 it works lol?

image


In 3.10, the "User Function" always shows `Unknown` for the type hint but not in 3.9

image

I'm seeing the same thing for the READMEs. In 3.9, the type hints get shown in the parameter table but not in 3.10.

File: gradio/cli/commands/components/_docs_utils.py
Comment (pngwn) [reply gradio/cli/commands/components/_docs_utils.py:34]: How are you switching your python version. `Line-length` _is` a valid parameter. I'm guessing that your ruff version for that python version is earlier than 0.1.4 when the param was added or maybe even prior to 0.1.2 when the formatter was added.

I've added a recent version of ruff as a dep now, so this shouldn't be an issue.

**Comment by pngwn** [comment]
@freddyaboulton I think the recent update should address everything, I've also added some prettier output and warnings/ errors for certain cases.

- If a user is on 3.9 or below then they will get a gentle warning at the end of the run that updating python will give them a better experience.
- If there is no `if __name__ == '__main__'"` then the docs command errors but tells the user how to fix the problem, or how to disable the check (`--suppress-demo-check`)

I think CI should be green now.

**Comment by pngwn** [comment]
I have a type error in 3.8 (i thought 3.9 was out min version?). I'll see if i can figure that one out.

**Comment by freddyaboulton** [comment]
Awesome @pngwn ! Can confirm that all of the bugs noted yesterday are now gone. Only have some questions around the content now.

* Capitalize the header titles

image

* Kind of confused by the User function section. It says `As input` for Annotated image but Annotated image is really only used an output component? I think we should explain this section before going into the `As input`

image

* ParamViewer is the only component that has an `As output` section. I think that's a bug?

* Some components do not have a `User function` section (video, annotated image, model3d, etc). If it's expected, can we omit the section? Without context it's a bit confusing what this means. 

image

**Comment by freddyaboulton** [comment]
One tiny bug is that you cannot generate docs for `gr.Button`. Says `TypeError: Object of type type is not JSON serializable`

**Comment by pngwn** [comment]
> Kind of confused by the User function section. It says As input for Annotated image but Annotated image is really only used an output component?

That isn't really true in a blocks world. It might be the intention but all component can be used as inputs or outputs, there is no way to restrict this so they documentation is correct and I'd think we should focus on explaining to authors that components should work as an input or an output even if they don't do much.

> ParamViewer is the only component that has an As output section. I think that's a bug?

I'll check how I'm pulling in the docstrings. We need a docstring for the return of preprocess and the param of postprocess to generate a meaningful entry.

> Some components do not have a User function section (video, annotated image, model3d, etc). If it's expected, can we omit the section? Without context it's a bit confusing what this means.

It's better than nothing imo, this happens because the params / returns  of processors are typed but there is no docstring (or there is a bug in pulling it in). We could use the type when there is no docstring to provide _some_ context.

I'll take a look at Button.

**Comment by freddyaboulton** [comment]
> We need a docstring for the return of preprocess and the param of postprocess to generate a meaningful entry.

Ok I think there may be a bug then, annotatedimage does not have docstrings for either pre or postprocess and it has an `As input` entry. Video does not have docstrings as well and it has no entry.

Regarding the `User function` section, I think we need to add a brief sentence or two explaining what it's trying to communicate and in what context it's relevant. Right now I find it a bit confusing.

File: gradio/components/paramviewer.py
Comment (freddyaboulton) [inline gradio/components/paramviewer.py:37]: Typo: `width syntax highlighting`

**Review by freddyaboulton — COMMENTED** [review]
Just noticed these two small things!

File: gradio/components/paramviewer.py
Comment (pngwn) [reply gradio/components/paramviewer.py:11]: Oh sorry, I'll remove these before merging.

**Comment by pngwn** [comment]
The reason the input/ output text was mixed up is because i got them mixed up. Annotated image docs should behave as expected now. Button thing was a bug and should be fixed. 

Just gonna fix that type error in 3.8 then I'll clean up the ParamViewer docs.

**Review by freddyaboulton — APPROVED** [review]
Thanks @pngwn !

Only thing I noticed is that the markdown should also have the helpful explanation of the User Function section. Otherwise LGTM once CI is fixed. Thanks!

image

**Comment by pngwn** [comment]
Ah yes ofc. Thanks for catching that!

**Pull request merged**
